src/memory_limit.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright 2017, 2018 Parity Technologies
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{Decode, Error, Input};
+
+/// The error message returned when the memory limit is reached.
+const DECODE_OOM_ERROR: &str = "Out of memory when decoding";
+
+/// Extension trait to [`Decode`] for decoding with a maximum memory consumption.
+///
+/// Should be used as a marker for types that do the memory tracking in their `decode` implementation with [`Input::try_alloc`].
+pub trait DecodeMemLimit: Decode + Sized {
+	fn decode_with_mem_limit<I: Input>(limit: MemLimit, input: &mut I) -> Result<Self, Error> {
+		let mut input = MemTrackingInput { inner: input, limit };
+		Self::decode(&mut input)
+	}
+}
+
+// Mark tuples as memory-tracking compatible if all of their elements are, since they dont consume any memory themselves.
+#[impl_trait_for_tuples::impl_for_tuples(18)]
+impl DecodeMemLimit for Tuple {
+	for_tuples!( where #( Tuple: DecodeMemLimit )* );
+
+	fn decode_with_mem_limit<I: Input>(limit: MemLimit, input: &mut I) -> Result<Self, Error> {
+		let mut input = MemTrackingInput { inner: input, limit };
+		let r = for_tuples!( ( #( Tuple::decode(&mut input)? ),* ) );
+		Ok(r)
+	}
+}
+
+/// An input that additionally tracks memory usage.
+pub struct MemTrackingInput<'a, I> {
+	/// The actual input.
+	pub inner: &'a mut I,
+
+	/// The remaining memory limit.
+	pub limit: MemLimit,
+}
+
+/// A limit on allocated memory.
+pub struct MemLimit {
+	/// The remaining memory limit.
+	limit: usize,
+	/// Memory alignment to be applied before allocating memory.
+	align: Option<MemAlignment>,
+}
+
+impl MemLimit {
+	/// Try to allocate a contiguous chunk of memory.
+	pub fn try_alloc(&mut self, size: usize) -> Result<(), Error> {
+		let size = self.align.as_ref().map_or(size, |a| a.align(size));
+
+		if let Some(remaining) = self.limit.checked_sub(size) {
+			self.limit = remaining;
+			Ok(())
+		} else {
+			Err(DECODE_OOM_ERROR.into())
+		}
+	}
+
+	/// Maximal possible limit.
+	pub fn max() -> Self {
+		Self { limit: usize::MAX, align: None }
+	}
+}
+
+/// Alignment of some amount of memory.
+///
+/// Normally the word `alignment` is used in the context of a pointer - not an amount of memory, but this is still
+/// the most fitting name.
+pub enum MemAlignment {
+	/// Round up to the next power of two.
+	NextPowerOfTwo,
```

Comment:
Note: Currently when using this with a Vector, we align each element individually.  
But a Vec allocates all elements in a chunk through the WASM bump allocator. So what we have to do is `round(sum(e...))` over all elements instead of `sum(round(e)..)`.  

Otherwise it would round each element size up to the next power of two, but the vector may still allocate more since it aligns with the sum of the sizes of all elements.